# Create ISA-API Investigation from Datascriptor Study Design configuration
# Crossover Study with radiological and drug treatment on Mice

In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.

Or study design configuration consists of:
- a 4-arm study design. Each arm has 10 subjects
- subjects are mice. There is an observational factor, named "sex" with two values: "M" and "F"
- a crossover of two treatments, a drug treatment (cyclophosphamide 500 mg for 20 days) and a radiological treatment ("strontium chloride sr 89" 2 MBq/Kg forr 40 days)
- four non-treatment phases: screen, run-in, washout and follow-up
- three sample types colllected: tissue, blood, and urine
- three assay types: 
    - metabolite profiling through mass spectrometry on the urine sample.
    - clinical chemistry with marker on blood samples
    - genomic sequencing on tissue samples

## 1. Setup

Let's import all the required libraries

In [1]:
from time import time
import os
import json

In [2]:
## ISA-API related imports
from isatools.model import Investigation, Study

In [3]:
## ISA-API create mode related imports
from isatools.create.model import StudyDesign
from isatools.create.connectors import generate_study_design

# serializer from ISA Investigation to JSON
from isatools.isajson import ISAJSONEncoder

# ISA-Tab serialisation
from isatools import isatab

In [4]:
## ISA-API create mode related imports
from isatools.create import model
from isatools import isajson

## 2. Load the Study Design JSON configuration

First of all we load the study design configurator with all the specs defined above

In [5]:
with open(os.path.abspath(os.path.join(
    "config", "crossover-study-design-2treatments-mice.json"
)), "r") as config_file:
    study_design_config = json.load(config_file)
study_design_config

{'name': 'Test study',
 'description': 'This design describes a study with repeated treatments on the same arms. We will be collecting three samples and these will undergo three different assays.',
 'subjectType': 'Mouse',
 'subjectSize': 10,
 'designType': {'term': 'crossover design',
  'id': 'OBI:0500003',
  'iri': 'http://purl.obolibrary.org/obo/OBI_0500003',
  'label': 'Study subjects receive repeated treatments',
  'value': 'crossover'},
 'observationalFactors': [{'name': 'sex',
   'values': ['M', 'F'],
   'isQuantitative': False,
   'unit': None}],
 'subjectGroups': {'selected': [{'name': 'SubjectGroup_0',
    'type': 'Mouse',
    'characteristics': [{'name': 'sex',
      'value': 'M',
      'unit': None,
      'isQuantitative': False}]},
   {'name': 'SubjectGroup_1',
    'type': 'Mouse',
    'characteristics': [{'name': 'sex',
      'value': 'F',
      'unit': None,
      'isQuantitative': False}]}],
  'unselected': []},
 'treatmentPlan': {'screen': {'selected': True,
   'name':

## 3. Generate the ISA Study Design from the JSON configuration
To perform the conversion we just need to use the function `generate_isa_study_design()` (name possibly subject to change, should we drop the "isa" and "datascriptor" qualifiers?)

In [6]:
study_design = generate_study_design(study_design_config)
assert isinstance(study_design, StudyDesign)

## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation

The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object.

In [7]:
start = time()
study = study_design.generate_isa_study()
end = time()
print('The generation of the study design took {:.2f} s.'.format(end - start))
assert isinstance(study, Study)
investigation = Investigation(identifier='inv01', studies=[study])

The generation of the study design took 1.32 s.


## 5. Serialize and save the JSON representation of the generated ISA Investigation

In [8]:
start = time()
inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))
end = time()
print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))

The JSON serialisation of the ISA investigation took 0.56 s.


In [9]:
directory = os.path.abspath(os.path.join('output', 'crossover-2treatments-mice'))
os.makedirs(directory, exist_ok=True)
with open(os.path.abspath(os.path.join(directory, 'isa-investigation-crossover-2treatments-mice.json')), 'w') as out_fp:
    json.dump(json.loads(inv_json), out_fp)

## 6. Dump the ISA Investigation to ISA-Tab

In [10]:
start = time()
isatab.dump(investigation, directory)
end = time()
print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))

The Tab serialisation of the ISA investigation took 25.29 s.


To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump

In [11]:
dataframes = isatab.dump_tables_to_dataframes(investigation)

In [12]:
len(dataframes)

4

## 7. Check the correctness of the ISA-Tab DataFrames 

We have 1 study file and 2 assay files (one for MS and one for NMR). Let's check the names:

In [13]:
for key in dataframes.keys():
    display(key)

's_study_01.txt'

'a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'

'a_AT11_clinical-chemistry_marker-panel.txt'

'a_AT1_metabolite-profiling_mass-spectrometry.txt'

### 7.1 Count of subjects and samples

We have 10 subjects in the each of the six arms for a total of 60 subjects. 5 blood samples per subject are collected (1 in treatment 1 phase, 1 in treatment, and 3 in the follow-up phase) for a total of 300 blood samples. These will undergo the NMR assay. We have 4 saliva samples per subject (1 during screen and 3 during follow-up) for a total of 240 saliva samples. These will undergo the "mass spcetrometry" assay.

In [14]:
study_frame = dataframes['s_study_01.txt']
count_arm0_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP0' in el)])
count_arm1_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP1' in el)])
count_arm2_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP2' in el)])
count_arm3_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP3' in el)])
print("There are {} samples in the GRP0 arm (i.e. group)".format(count_arm0_samples))
print("There are {} samples in the GRP1 arm (i.e. group)".format(count_arm1_samples))
print("There are {} samples in the GRP2 arm (i.e. group)".format(count_arm2_samples))
print("There are {} samples in the GRP3 arm (i.e. group)".format(count_arm3_samples))

There are 160 samples in the GRP0 arm (i.e. group)
There are 160 samples in the GRP1 arm (i.e. group)
There are 160 samples in the GRP2 arm (i.e. group)
There are 160 samples in the GRP3 arm (i.e. group)


In [20]:
study_frame

Source Name Characteristics[Study Subject] Characteristics[sex]  \
0    GRP0_SBJ01                          Mouse                    M   
1    GRP0_SBJ01                          Mouse                    M   
2    GRP0_SBJ01                          Mouse                    M   
3    GRP0_SBJ01                          Mouse                    M   
4    GRP0_SBJ01                          Mouse                    M   
..          ...                            ...                  ...   
635  GRP3_SBJ10                          Mouse                    F   
636  GRP3_SBJ10                          Mouse                    F   
637  GRP3_SBJ10                          Mouse                    F   
638  GRP3_SBJ10                          Mouse                    F   
639  GRP3_SBJ10                          Mouse                    F   

          Protocol REF Parameter Value[run order] Parameter Value[study cell]  \
0    sample collection                        112                        A0E5   
1    sample collection                        133                        A0E5   
2    sample collection                        001                        A0E0   
3    sample collection                        041                        A0E2   
4    sample collection                        111                        A0E5   
..                 ...                        ...                         ...   
635  sample collection                        493                        A3E0   
636  sample collection                        594                        A3E5   
637  sample collection                        615                        A3E5   
638  sample collection                        494                        A3E0   
639  sample collection                        593                        A3E5   

           Date Performer                         Sample Name  \
0    2021-03-11       Bob         GRP0_SBJ01_A0E5_SMP-Urine-2   
1    2021-03-11       Bob  GRP0_SBJ01_A0E5_SMP-Blood-Sample-3   
2    2021-03-11       Bob  GRP0_SBJ01_A0E0_SMP-Blood-Sample-1   
3    2021-03-11       Bob  GRP0_SBJ01_A0E2_SMP-Blood-Sample-1   
4    2021-03-11       Bob         GRP0_SBJ01_A0E5_SMP-Urine-1   
..          ...       ...                                 ...   
635  2021-03-11       Bob         GRP3_SBJ10_A3E0_SMP-Urine-1   
636  2021-03-11       Bob         GRP3_SBJ10_A3E5_SMP-Urine-2   
637  2021-03-11       Bob  GRP3_SBJ10_A3E5_SMP-Blood-Sample-2   
638  2021-03-11       Bob         GRP3_SBJ10_A3E0_SMP-Urine-2   
639  2021-03-11       Bob         GRP3_SBJ10_A3E5_SMP-Urine-1   

    Characteristics[organism part] Factor Value[DURATION]  Unit  \
0                            Urine                     90  days   
1                     Blood Sample                     90  days   
2                     Blood Sample                     10  days   
3                     Blood Sample                     40  days   
4                            Urine                     90  days   
..                             ...                    ...   ...   
635                          Urine                     10  days   
636                          Urine                     90  days   
637                   Blood Sample                     90  days   
638                          Urine                     10  days   
639                          Urine                     90  days   

          Factor Value[AGENT] Factor Value[INTENSITY]  Unit.1  
0                                                              
1                                                              
2                                                              
3    strontium chloride sr 89                     2.0  MBq/Kg  
4                                                              
..                        ...                     ...     ...  
635                                                            
636                                                            
637             

In [15]:
dataframes['a_AT1_metabolite-profiling_mass-spectrometry.txt']

Sample Name Protocol REF Performer         Extract Name  \
0     GRP0_SBJ01_A0E0_SMP-Urine-1   extraction   Unknown    AT1-S1-Extract-R1   
1     GRP0_SBJ01_A0E0_SMP-Urine-1   extraction   Unknown    AT1-S1-Extract-R1   
2     GRP0_SBJ01_A0E0_SMP-Urine-1   extraction   Unknown    AT1-S1-Extract-R1   
3     GRP0_SBJ01_A0E0_SMP-Urine-1   extraction   Unknown    AT1-S1-Extract-R1   
4     GRP0_SBJ01_A0E0_SMP-Urine-2   extraction   Unknown    AT1-S2-Extract-R1   
...                           ...          ...       ...                  ...   
1115  GRP3_SBJ10_A3E5_SMP-Urine-1   extraction   Unknown  AT1-S263-Extract-R1   
1116  GRP3_SBJ10_A3E5_SMP-Urine-2   extraction   Unknown  AT1-S264-Extract-R1   
1117  GRP3_SBJ10_A3E5_SMP-Urine-2   extraction   Unknown  AT1-S264-Extract-R1   
1118  GRP3_SBJ10_A3E5_SMP-Urine-2   extraction   Unknown  AT1-S264-Extract-R1   
1119  GRP3_SBJ10_A3E5_SMP-Urine-2   extraction   Unknown  AT1-S264-Extract-R1   

     Characteristics[extract type] Term Accession Number Protocol REF.1  \
0                   polar fraction        polar fraction       labeling   
1                   polar fraction        polar fraction       labeling   
2                   polar fraction        polar fraction       labeling   
3                   polar fraction        polar fraction       labeling   
4                   polar fraction        polar fraction       labeling   
...                            ...                   ...            ...   
1115                polar fraction        polar fraction       labeling   
1116                polar fraction        polar fraction       labeling   
1117                polar fraction        polar fraction       labeling   
1118                polar fraction        polar fraction       labeling   
1119                polar fraction        polar fraction       labeling   

     Performer.1 Labeled Extract Name    Label     Protocol REF.2  \
0        Unknown         AT1-S1-LE-R1  label_0  mass spectrometry   
1        Unknown         AT1-S1-LE-R1  label_0  mass spectrometry   
2        Unknown         AT1-S1-LE-R1  label_0  mass spectrometry   
3        Unknown         AT1-S1-LE-R1  label_0  mass spectrometry   
4        Unknown         AT1-S2-LE-R1  label_0  mass spectrometry   
...          ...                  ...      ...                ...   
1115     Unknown       AT1-S263-LE-R1  label_0  mass spectrometry   
1116     Unknown       AT1-S264-LE-R1  label_0  mass spectrometry   
1117     Unknown       AT1-S264-LE-R1  label_0  mass spectrometry   
1118     Unknown       AT1-S264-LE-R1  label_0  mass spectrometry   
1119     Unknown       AT1-S264-LE-R1  label_0  mass spectrometry   

     Parameter Value[instrument]                    Term Accession Number.1  \
0              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
1              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
2              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
3              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
4              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
...                          ...                                        ...   
1115           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
1116           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
1117           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
1118           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   
1119           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   

     Parameter Value[injection_mode]  \
0                                 LC   
1                                FIA   
2                                 LC   
3                                FIA   
4                                FIA   
...                              ...   
1115                              LC   
1116                             FIA   
1117      

In [16]:
dataframes['a_AT11_clinical-chemistry_marker-panel.txt']

Sample Name        Protocol REF Performer  \
0    GRP0_SBJ01_A0E0_SMP-Blood-Sample-1  sample preparation   Unknown   
1    GRP0_SBJ01_A0E0_SMP-Blood-Sample-1  sample preparation   Unknown   
2    GRP0_SBJ01_A0E0_SMP-Blood-Sample-1  sample preparation   Unknown   
3    GRP0_SBJ01_A0E2_SMP-Blood-Sample-1  sample preparation   Unknown   
4    GRP0_SBJ01_A0E2_SMP-Blood-Sample-1  sample preparation   Unknown   
..                                  ...                 ...       ...   
955  GRP3_SBJ10_A3E5_SMP-Blood-Sample-2  sample preparation   Unknown   
956  GRP3_SBJ10_A3E5_SMP-Blood-Sample-2  sample preparation   Unknown   
957  GRP3_SBJ10_A3E5_SMP-Blood-Sample-3  sample preparation   Unknown   
958  GRP3_SBJ10_A3E5_SMP-Blood-Sample-3  sample preparation   Unknown   
959  GRP3_SBJ10_A3E5_SMP-Blood-Sample-3  sample preparation   Unknown   

                  Raw Data File  
0      AT11-S2-raw_data_file-R1  
1      AT11-S3-raw_data_file-R1  
2      AT11-S1-raw_data_file-R1  
3     AT11-S31-raw_data_file-R1  
4     AT11-S32-raw_data_file-R1  
..                          ...  
955  AT11-S884-raw_data_file-R1  
956  AT11-S883-raw_data_file-R1  
957  AT11-S888-raw_data_file-R1  
958  AT11-S886-raw_data_file-R1  
959  AT11-S887-raw_data_file-R1  

[960 rows x 4 columns]

In [17]:
dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt']

Sample Name Protocol REF Performer  \
0    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1   extraction   Unknown   
1    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1   extraction   Unknown   
2    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1   extraction   Unknown   
3    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1   extraction   Unknown   
4    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1   extraction   Unknown   
..                                   ...          ...       ...   
635  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1   extraction   Unknown   
636  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1   extraction   Unknown   
637  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1   extraction   Unknown   
638  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1   extraction   Unknown   
639  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1   extraction   Unknown   

           Extract Name Characteristics[extract type]       Protocol REF.1  \
0     AT8-S2-Extract-R2                          gDNA  library_preparation   
1     AT8-S1-Extract-R2                          gDNA  library_preparation   
2     AT8-S2-Extract-R2                          gDNA  library_preparation   
3     AT8-S1-Extract-R2                          gDNA  library_preparation   
4     AT8-S1-Extract-R2                          gDNA  library_preparation   
..                  ...                           ...                  ...   
635  AT8-S64-Extract-R1                           DNA  library_preparation   
636  AT8-S64-Extract-R1                           DNA  library_preparation   
637  AT8-S64-Extract-R1                           DNA  library_preparation   
638  AT8-S63-Extract-R2                          gDNA  library_preparation   
639  AT8-S63-Extract-R1                           DNA  library_preparation   

    Parameter Value[instrument] Parameter Value[library_orientation]  \
0         Illumina NovaSeq 6000                               single   
1         Illumina NovaSeq 6000                               single   
2         Illumina NovaSeq 6000                               single   
3         Illumina NovaSeq 6000                               single   
4         Illumina NovaSeq 6000                               single   
..                          ...                                  ...   
635       Illumina NovaSeq 6000                               single   
636       Illumina NovaSeq 6000                               single   
637       Illumina NovaSeq 6000                               single   
638       Illumina NovaSeq 6000                               single   
639       Illumina NovaSeq 6000                               single   

    Parameter Value[library_strategy] Performer.1  \
0                                 WGS     Unknown   
1                            AMPLICON     Unknown   
2                                 WGS     Unknown   
3                                 WGS     Unknown   
4                            AMPLICON     Unknown   
..                                ...         ...   
635                          AMPLICON     Unknown   
636                               WGS     Unknown   
637                          AMPLICON     Unknown   
638                               WGS     Unknown   
639                               WGS     Unknown   

                    Raw Data File  
0     AT8-S2-raw_data_file-R5.raw  
1     AT8-S1-raw_data_file-R8.raw  
2     AT8-S2-raw_data_file-R6.raw  
3     AT8-S1-raw_data_file-R6.raw  
4     AT8-S1-raw_data_file-R7.raw  
..                            ...  
635  AT8-S64-raw_data_file-R2.raw  
636  AT8-S64-raw_data_file-R3.raw  
637  AT8-S64-raw_data_file-R1.raw  
638  AT8-S63-raw_data_file-R6.raw  
639  AT8-S63-raw_data_file-R4.raw  

[640 rows x 11 columns]

# 7.1 Overview of the Mass Spec assay table

For the mass. spec. assay table, we have 160 (derma) samples, 160 extracts (1 per  sample, "polar" fraction), 320 labeled extracts (2 per extract, as "#replicates" is  2) and 640 mass spec protocols + 640 output files (2 per labeled extract as we do 1 technical replicate with 2 protocol parameter combinations `["Agilent QTQF 6510", "FIA", "positive mode"]` and `["Agilent QTQF 6510", "LC", "positive mode"]`).

In [18]:
dataframes['a_AT1_metabolite-profiling_mass-spectrometry.txt'].nunique(axis=0, dropna=True)

Sample Name                           280
Protocol REF                            1
Performer                               1
Extract Name                          280
Characteristics[extract type]           1
Term Accession Number                   1
Protocol REF.1                          1
Performer.1                             1
Labeled Extract Name                  280
Label                                   1
Protocol REF.2                          1
Parameter Value[instrument]             1
Term Accession Number.1                 1
Parameter Value[injection_mode]         2
Term Accession Number.2                 2
Parameter Value[acquisition_mode]       1
Term Accession Number.3                 1
MS Assay Name                        1120
Performer.2                             1
Raw Spectral Data File               1120
dtype: int64

### Overview of the Genome Sequencing assay table

For the NMR assay table, we have 360 (blood) samples, 720 extracts (2 per  sample, a single replicate of the "supernatant" and "pellet" fractions) and 2880 NMR protocols + 4800 output files (4 per extract as we do 2 technical replicates with 2 protocol parameter combinations `["Bruker Avance II 1 GHz", "1D 1H NMR", "CPGM"]` and `["Bruker Avance II 1 GHz", "1D 1H NMR", "TOCSY"]`).

In [19]:
dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'].nunique(axis=0, dropna=True)

Sample Name                              40
Protocol REF                              1
Performer                                 1
Extract Name                            160
Characteristics[extract type]             2
Protocol REF.1                            1
Parameter Value[instrument]               1
Parameter Value[library_orientation]      1
Parameter Value[library_strategy]         2
Performer.1                               1
Raw Data File                           640
dtype: int64